add pybind version of HANDLE_TH_ERRORS

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27154 Handle warning in torchscript
* **#26614 add pybind version of HANDLE_TH_ERRORS**
* #26613 Buffer python warning to avoid deadlocks
* #28283 Fix clang-tidy errors in csrc/Module.cpp

Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

albanD committed:
add pybind version of HANDLE_TH_ERRORS

[ghstack-poisoned]

albanD committed:
Update on "[WIP] add pybind version of HANDLE_TH_ERRORS"




[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

ezyang commented on test/test_torch.py:
```diff
@@ -6892,6 +6893,86 @@ def test_inverse(self, device):
+        // 2: python_error()
+        // 3: py::error_already_set
+        at::Tensor catched(at::Tensor x, int error_type) {
+            HANDLE_TH_ERRORS
```
Heh, we call it `HANDLE_TH_ERRORS` in our codebase because legacy mumble; no reason to call it that here. `TORCH_HANDLE_ERRORS` maybe?

albanD replied:
As described in details in the doc, the new macro is not superseeding the old one, it is just used in a different context.
I would be in favor of keeping the same name base here to specify that it is just a specialization of the original macro for pybind-bound functions.
But if you really think we should change, I can do it.

ezyang reviewed (changes requested):
This is going to need documentation updates

ezyang commented:
(The substantive content of the PR looks good though!)

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

Commit:
Update on "[WIP] add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

Commit:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D17692503](https://our.internmc.facebook.com/intern/diff/D17692503)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

albanD committed:
Update on "add pybind version of HANDLE_TH_ERRORS"


Differential Revision: [D18249634](https://our.internmc.facebook.com/intern/diff/D18249634)

[ghstack-poisoned]

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -102,6 +102,25 @@ def _find_cuda_home():
 
+# See comment in load_inline for more information
+# The goal is to be able to call the safe version of the
+# function exactely as if it was the original one.
```
nit: exactly

ezyang commented on torch/csrc/README.md:
```diff
@@ -68,6 +75,27 @@ exception which doesn't contain any info, instead it says, "An error
+  TORCH_CHECK(cond, "cond was false here");
+  TORCH_WARN("Warning message");
+  ...
+  END_HANDLE_TH_ERRORS_PYBIND
```
Sigh, it's gonna be so hard to get people to actually use this macro XD

albanD replied:
We could provide a `get_safe_version` like the one used in `load_inline` in the whole code base and recommend all pybind bindings to use `get_safe_version(foo)` instead of `foo` in their `m.def` ?
